Add Billing Wrapper

itatulgarg opened the pull request:
initial version of the Billing Wrapper DLL for .NET clients

itatulgarg committed:
Add Billing Wrapper

initial version of the Billing Wrapper DLL for .NET clients

mattcarrier commented on billingWrapper/Intersections.Billing/BillingToken.cs:
```diff
@@ -0,0 +1,71 @@
+        /// </summary>
+        /// <param name="tokenRequest">TokenRequest containing billing information</param>
+        /// <returns>TokenResponse</returns>
+        public async Task<TokenResponse> get(TokenRequest tokenRequest, string apiKey)
```
are we expecting them to give us the token?  I would expect us to have the token directly in this wrapper so the end-client doesn't even have to worry about it.  I mean right now it's visible in the UI source anyway so it's basically public info at this point.

itatulgarg replied:
They will give not give us the Token, rather they will provide us with TokenRequest consisting of Billing Address and Card Info etc. In return they will get the TokenResponse consisting of either the token or errors as the case may be.

mattcarrier replied:
oh man I guess I wrote that question too quickly (pardon me), I meant the apiKey.  You have a direct parameter that accepts apiKey here so that means that we need a process in place to give them the apikey so they can pass it in.  As well as a process to send them updates if we need to change the apikey.  It seems to me that this would be an implementation of the wrapper, since it is directly related to the billing vendor.  If we change vendors our apiKey would must definitely change.  My thoughts would be to encapsulate it.

mattcarrier commented:
generally I think this is great.  I was wrestling with the fact whether we want the test source in there but I think it's good for the partners to see actual uses of the wrapper.  My only real question/concern is that I want to make sure we're not relying on the partner to give us the API key.  We should have that encapsulated in our wrapper so they only have to give us the request details and that's it.

itatulgarg commented:
The API key is the Recurly API Key which varies based on the environment. Since the environment varies when developing/testing/staging/prod etc., we need a way to specify the key before calling recurly endpoint. Other way to achieve this would be to have the client code specify the environment while calling the library function which I thought would be a little odd. Thoughts?

mattcarrier commented:
@itatulgarg That is a really good point!  We could have a parameter that they pass in to say whether it is "pre-prod" or "prod".  But really at the end of the day it's on the vendor to tell us whether they are hitting our "pre-prod" billing instance or "prod" billing instance.   I'm assuming most vendors have an available variable/method that tells them what environment they are in, so I'm assuming that wouldn't be too much work for them.

And this really just escalates the problem that we don't want to have our vendors manage our api keys.  We now have 2 api keys that they would have to store and we have to keep some kind of mapping of which vendors have what keys and which ones we need to send the new key to.  It just sounds like quite a nightmare to me.

itatulgarg commented:
I agree 100%

itatulgarg committed:
remove apiKey from client call

Added a flag for the client code to specify whether its a TEST environment. If the value is true, the library will pick the STAGING key for recurly, otherwise it will pick PROD key

itatulgarg commented:
@mattcarrier updated as discussed, please review

itatulgarg committed:
Create README.md

itatulgarg committed:
Update README.md

mattcarrier commented:
@itatulgarg Everything looks good.  The only thing I would like to see is that instead of 'isTest' we call it 'isPreProd'.  If you agree with that change once you do that feel free to merge the PR.  If you don't agree with that then let's have a convo.

itatulgarg committed:
updated as per review
